### PR TITLE
Update sanctuary

### PR DESCRIPTION
**Pull request by dotnetCarpenter**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:
If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
  -  https://sanctuary.js.org/#size
  -  https://sanctuary.js.org/#fromPairs
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/sanctuary/index.d.ts [hunk 1]
```diff
@@ -13,38 +14,40 @@ export as namespace S;
...
-type Fn5_<A, B, C, D, E, F>  = (a: A, b: B, c: C, d: D, e: E) => F;
+type Fn<A, B>               = (a: A) => B;
+type Fn2<A, B, C>           = (a: A) => (b: B) => C;
+type Fn3<A, B, C, D>        = (a: A) => (b: B) => (c: C) => D;
+type Fn4<A, B, C, D, E>     = (a: A) => (b: B) => (c: C) => (d: D) => E;
+type Fn5<A, B, C, D, E, F>  = (a: A) => (b: B) => (c: C) => (d: D) => (e: E) => F;
+type Fn2_<A, B, C>          = (a: A, b: B) => C;
+type Fn3_<A, B, C, D>       = (a: A, b: B, c: C) => D;
+type Fn4_<A, B, C, D, E>    = (a: A, b: B, c: C, d: D) => E;
+type Fn5_<A, B, C, D, E, F> = (a: A, b: B, c: C, d: D, e: E) => F;
```
Comment (davidchambers) [inline types/sanctuary/index.d.ts:30]: Could you revert these changes and the other formatting changes to make the significant changes easier to review? :)

Comment (dotnetCarpenter) [reply types/sanctuary/index.d.ts]: I dropped the commit that used DefinitelyTyped's prettier. Is that better @davidchambers?

File: types/sanctuary/index.d.ts [hunk 2]
```diff
@@ -48,28 +49,31 @@ type NonNegativeInteger     = number;
...
+interface Bifunctor<A, C> extends Functor<C> {}
+interface Category<A> extends Semigroupoid<A, A> {}
 interface Chain<A> extends Apply<A> {}
 interface ChainRec<A> extends Chain<A> {}
+interface Comonad<A> extends Extend<A> {}
+interface Contravariant<A> {}
+interface Extend<A> extends Functor<A> {}
+interface Foldable<A> {}
+interface Functor<A> {}
+interface Group<A> extends Monoid<A> {
```
Comment (dotnetCarpenter) [inline types/sanctuary/index.d.ts:65]: @davidchambers `Group<A>` is the only new Type Class

Comment (dotnetCarpenter) [reply types/sanctuary/index.d.ts:65]: Used for [`S.invert`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59923/files#diff-5f5b52c902d2832b88260d5f71d8957cd9fd861a6cc3b4a65b5dc92c4af08b9fR120)

**Comment by dotnetCarpenter** [comment]
@davidchambers This PR adds new type definition to the following 7 functions:

1. `reduce_`
2. `invert`
3. `size`
4. `singleton`
5. `insert`
6. `remove`
7. `fromPairs`

This PR updates the type definition of the following function:

1. `value` https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59923/commits/9e92bc573be1a8ddf3afe636fad136e5aa65886e

This PR adds new tests of type definition for the following 12 functions:

1. `reduce`
2. `reduce_`
3. `invert`
4. `size`
5. `value`
6. `singleton`
7. `insert`
8. `remove`
10. `keys`
11. `values`
12. `pairs`
13. `fromPairs`

**Comment by dotnetCarpenter** [comment]
@davidchambers So one cool thing I found, [if you use curried functions, then typescript can infer the argument](https://github.com/microsoft/TypeScript/issues/26242#issuecomment-895295269) - probably not a great way to describe it but see the example in the link.

As an example from this PR, take the `value` function.

It used to be:
```ts
interface value<A>(p: string): (p: StrMap<A>) => Maybe<A>;

// which mean that the return type `Maybe<A>` had to be explicitly set
// on `S.value` according to the `StrMap`.

// $ExpectType Maybe<number>
S.value<number> ('foo') ({foo: 1, bar: 2});

// $ExpectType Maybe<string>
S.value<string> ('foo') ({foo: 'hallo', bar: 'world'});
```

But if we change `value`'s definition to:

```ts
interface value(p: string): <A>(p: StrMap<A>) => Maybe<A>;

// then typescript can infer what type `A` is.

// $ExpectType Maybe<number>
S.value ('foo') ({foo: 1, bar: 2});

// $ExpectType Maybe<string>
S.value ('foo') ({foo: 'hallo', bar: 'world'});
```

Now, there are other ways to help typescript to infer types. For example with `S.reduce`, typescript needs to know the types of the function parameters before it can establish the return value.

```ts
interface reduce<A, B>(p: Fn2<B, A, B>): (q: B) => (r: ReadonlyArray<A> | StrMap<A> | Maybe<A> | Either<any, A> | Foldable<A>) => B;

// $ExpectType number[]
S.reduce<number, number[]> (xs => x => S.prepend (x) (xs))
                           ([])
                           ([1, 2, 3, 4, 5]);
// [ 5, 4, 3, 2, 1 ]
```

Without knowing `<A, B>` typescript can not say anything meaningful about `x` or the return type. But if you give it a function with well defined parameters, then typescript can infer the data types.

```ts
interface add(p: FiniteNumber): (q: FiniteNumber) => FiniteNumber;

// $ExpectType number
S.reduce (S.add) (0) ([1, 2, 3, 4, 5]);
// 15
```

My point is, that there is probably more definitions where types can be inferred automatically with a little change to the type definition. But in a lot of cases, like `S.reduce`, I do not see how changing the definition can help. Rather, being more specific with the concrete data types seems to help.

E.g.
```ts
// $ExpectType number[]
S.reduce ((xs: number[]) => (x: number) => S.prepend (x) (xs))
         ([])
         ([1, 2, 3, 4, 5]);
```

I feel the same rationale should work for `S.chain` and friends.. If the type classes are better defined then they might work. Similar to how [Fluture define Functor and Monad](https://github.com/fluture-js/Fluture/blob/master/index.d.ts#L26). But it might be a lot of work and I'm not sure it will solve the problem with `Chain<unknown>`.

```
Type 'Chain<unknown>' is missing the following properties from type 'FutureInstance<string, string>': pipe, extractLeft, extractRight, 'fantasy-land/ap', and 5 more
```

I took the first baby step with [`Group<A>`](https://github.com/fantasyland/fantasy-land/tree/v4.0.1#group) and perhaps one could take another look at [using `type` instead of `interface`.](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45895#discussion_r449951036)

_I'm not using typescript directly but vscode so my issues might be conflated with the use of vscode and not necessarily typescript only. But now I'm intrigued!_

File: types/sanctuary/index.d.ts (hunk 2 above)
Comment (davidchambers) [reply types/sanctuary/index.d.ts:65]: Could we preserve the current ordering, and simply insert `Group<A>` after `Monoid<A>`?

Comment (dotnetCarpenter) [reply types/sanctuary/index.d.ts:65]: Sure but I'm curious to why you do not prefer alphabetic ordering? Ahh I see, you are using the ordering from https://github.com/sanctuary-js/sanctuary-type-classes#type-class-hierarchy, right?
I'll add a commit that we retain the original order.

File: types/sanctuary/index.d.ts [hunk 3]
```diff
@@ -48,12 +49,17 @@ type NonNegativeInteger     = number;
...
+// https://github.com/sanctuary-js/sanctuary-type-classes#type-class-hierarchy
 interface Setoid<A> {}
 interface Ord<A> extends Setoid<A> {}
 interface Semigroupoid<A, B> {}
 interface Category<A> extends Semigroupoid<A, A> {}
 interface Semigroup<A> {}
 interface Monoid<A> extends Semigroup<A> {}
+interface Group<A> extends Monoid<A> {
+  'fantasy-land/invert': () => Group<A>;
+}
```
Comment (davidchambers) [inline types/sanctuary/index.d.ts:62]: I think we should use `{}` here rather than `{ 'fantasy-land/invert': () => Group<A>; }`. I recall there being a reason we could not use the proper interfaces (built-in types, perhaps).

Comment (dotnetCarpenter) [reply types/sanctuary/index.d.ts:62]: If you look at [how Fluture defines its types](https://github.com/fluture-js/Fluture/blob/master/index.d.ts#L26), you can see that the fantasy-land specification are defined within `{}`. My implementation is possibly a little naive...

> I feel the same rationale should work for S.chain and friends.. If the type classes are better defined then they might work. Similar to how [Fluture define Functor and Monad](https://github.com/fluture-js/Fluture/blob/master/index.d.ts#L26).

If this prevents this PR for being merged then I will happily remove it until the implications are better understood. I guess the real test is to add the entire fantasy-land specification and see if typescript can transpile a demo app.. But unfortunately I do not have spare time to try that out at the moment.

Comment (davidchambers) [reply types/sanctuary/index.d.ts:62]: It makes sense to me to use `{}` for now, then submit a separate pull request to update all the interfaces. :)

**Review by davidchambers — APPROVED** [review]

**Comment by dotnetCarpenter** [comment]
@davidchambers seems that my excessive pushing has confused @typescript-bot. I'm not 100% sure but I think you need to approve again. Sorry

**Review by davidchambers — APPROVED** [review]

**Comment by dotnetCarpenter** [comment]
Ready to merge
